applications/ParticleMechanicsApplication/custom_constitutive/hencky_plastic_3D_law.cpp

```diff
@@ -189,7 +189,7 @@ void HenckyElasticPlastic3DLaw::InitializeMaterial(const Properties& rProps,
     bool restarted = false;
 
     if (rShapeFunctionsValues.size() > 0){
-        if (rShapeFunctionsValues(0) != -888 ){
+        if (rShapeFunctionsValues(0) == -888 ){
```

Comment:
I do not understand why you need to set anything related to the ShapeFunctions in the CL, can you develop a bit?

Reply:
@philbucher don't we have a `rProcessInfo[IS_RESTARTED]` capability?

Reply:
Yes. Indeed, this is the purposed it is intended to be used for.

Reply:
Yes however, I don't know how to access the ProcessInfo from within this function, as it is not passed in the function call and I could not find it being saved within the class (I might be wrong).

Reply:
True. I think you'll need to do it through the properties...

Reply:
Indeed, I can't understand why we are not passing the `ProcessInfo` in here. @KratosMultiphysics/technical-committee .

Reply:
Theoretically, the Initialize is done only once. If it is restarted it does not make sense to do it again (IMO). All the parameters to be initialized are serialized

Reply:
I thought so too, but if I don't reinitialize the `mpMPMFlowRule` in the hencky constitutive law the restart does not work. In the case of using a linear elastic law the whole initialize can be skipped, but for the hencky law it didn't work.

Reply:
I would pass the ProcessInfo.

Then it is consistent with other parts of the code and everyone can do what they need

To me this looks like a leftover

Reply:
Aside of the `ProcessInfo` discussion. This seems a bug in the serialization of one of the constitutive laws...

Reply:
True. I did not verify the serialization within the constitutive laws, but I will check if the `mpMPMFlowRule` is skipped and in that case if I can add it.

Reply:
I cannot come up with a solution on how to solve the serializer problem for the particle hardening law (I don't know if someone else had this problem with the recursive call before?). To be fair the information saved there only includes the material properties from the `Materials.json` file, so even when restarting these values can be imported again.
However, this means it needs to be initialized, which goes back to the initial problem of this strange workaround. To solve this workaround it's necessary to either pass the `rProcessInfo` to the `InitializeMaterial()` function or to include a new function for example `InitializeMaterialRestart()` in the `constitutive_law.h`. Maybe the second option does not make so much sense if this application is the only one that needs it, even though in my case it would be better since I would not need to include an if-statement in cases where the whole initialization can be skipped. Both of these solutions include modifying the core though.

Reply:
@philbucher @rubenzorrilla @AlejandroCornejo If you all agree I would open an issue assigning the technical committee to decide what is the best way to proceed. I would suggest just passing the `ProcessInfo` as it will be the easiest way to implement this and could be useful in the future as well.

Reply:
> This seems to be the problem here:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/057edcf143648d0ece0df3595d50bd8f0c44063a/applications/ParticleMechanicsApplication/custom_constitutive/hardening_laws/particle_hardening_law.hpp#L259-L273

**Ugh**, if that can't be solved, I'd at least do one of the following:
- throw an exception when `save` is called (best indicator of failure in my opinion: fails exactly when the problem arises)
- throw an exception when `load` is called (not great, not terrible: fails when your results are about to get messed up)
- issue warnings in `save` and `load` (suboptimal: noone cares about warnings until they've hunted the bug for a couple of days)

Reply:
The problem with throwing an exception is that the restart application would never work in this case. When the restart application tries to save or load this information it would then fail every time.
As @philbucher mentioned in the next comment, the other applications do not save these properties, but instead, pass the properties whenever necessary from the outside.
However, if this is to be changed, then all calls involving the `particle_hardening_law` would need to be changed.

Reply:
Well, if there's no way of omitting them during serialization/deserialization, all that's left is at least to issue a warning.

Reply:
By the way, I think it would be best to follow up on why serialization is not possible. The current behaviour is a bug and no amount of workarounds will change that. Furthermore, it might be straightforward for us right now to follow how the restarted initialization happens, but I guarantee that it won't be for others who take a look at it for the first time.

I haven't looked into it yet, but I have a sneaking suspicion that we'll find other bugs if we start looking into this more closely. Storing `shared_ptr`s to each other definitely smells like a memory leak.

Reply:
I agree, I will look into this. I will try to change the current implementation to avoid storing the material properties in this hardening_law class and instead pass the properties to all of the functions that need them. Maybe this way it would also not be necessary to have an `is_restarted` flag in the `InitializeMaterial()` function.

I will follow up on that.

Reply:
Ok so I deleted the attribute `mpProperties` inside the `particle_hardening_law` now and instead pass it as a parameter to all the functions that need it. This way the `mpProperties` do not need to be serialized anymore and thus avoid storing pointers to each other.
See PR #10667.